apiserver/facades/agent/uniter/uniter.go

```diff
@@ -1589,49 +1630,65 @@ func (u *UniterAPI) ReadRemoteSettings(ctx context.Context, args params.Relation
 	if err != nil {
 		return params.SettingsResults{}, err
 	}
+	for i, arg := range args.RelationUnitPairs {
+		settings, err := u.readOneRemoteSettings(ctx, canAccess, arg)
+		result.Results[i].Error = apiservererrors.ServerError(err)
+		result.Results[i].Settings = settings
+	}
+
+	return result, nil
+}
+
+func (u *UniterAPI) readOneRemoteSettings(ctx context.Context, canAccess common.AuthFunc, arg params.RelationUnitPair) (params.Settings, error) {
+	unitTag, err := names.ParseUnitTag(arg.LocalUnit)
+	if err != nil {
+		return nil, apiservererrors.ErrPerm
+	}
 
-	readOneSettings := func(arg params.RelationUnitPair) (params.Settings, error) {
-		unit, err := names.ParseUnitTag(arg.LocalUnit)
+	if !canAccess(unitTag) {
+		return nil, apiservererrors.ErrPerm
+	}
+
+	remoteTag, err := names.ParseTag(arg.RemoteUnit)
+	if err != nil {
+		return nil, apiservererrors.ErrPerm
```

Comment:
Same here, are these the correct errors in this case?

Reply:
Yes, this is what is done in the code already. I was surprised myself, but chose not to change the error expectations at this time.